File: src/jaslib_src/jaslib/language_services/janome_ex/word_extraction/matches/state_tests/head/failed_match_requirement.py
# from __future__ import annotations
#
# from typing import override
#
# from autoslot import Slots
#
# from jaslib.language_services.janome_ex.word_extraction.matches.requirements.requirement import MatchRequirement
#
#
# class FailedMatchRequirement(MatchRequirement, Slots):
#     def __init__(self, reason: str) -> None:
#         self.reason: str = reason
#
#     @property
#     @override
#     def failure_reason(self) -> str: return self.reason
#
#     @property
#     @override
#     def is_fulfilled(self) -> bool: return False
#
#     @classmethod
#     def forbids(cls, message: str) -> FailedMatchRequirement: return FailedMatchRequirement(f"forbids::{message}")
#
#     @classmethod
#     def required(cls, message: str) -> FailedMatchRequirement: return FailedMatchRequirement(f"required::{message}")

File: src/src_dotnet/JAStudio.Core/LanguageServices/JanomeEx/WordExtraction/Matches/StateTests/Head/FailedMatchRequirement.cs
using JAStudio.Core.LanguageServices.JanomeEx.WordExtraction.Matches.Requirements;

namespace JAStudio.Core.LanguageServices.JanomeEx.WordExtraction.Matches.StateTests.Head;

public sealed class FailedMatchRequirement : MatchRequirement
{
   public string Reason { get; }

   FailedMatchRequirement(string reason) => Reason = reason;

   public override string FailureReason => Reason;
   public override bool IsFulfilled => false;

   public static FailedMatchRequirement Forbids(string message) =>
      new($"forbids::{message}");

   public static FailedMatchRequirement Required(string message) =>
      new($"required::{message}");
}
